katello/katello-restore

```diff
@@ -31,39 +42,114 @@ end
+  `katello-service start --only #{necessary_services}`
+  `katello-service stop --only #{pulp_services}`
+  `sudo -u apache pulp-manage-db`
+  puts "Done."
+end
+
+def restore_mongo_dump
+  puts "Restoring mongo dump"
+  `katello-service start --only mongod`
+  `echo 'db.dropDatabase();' | mongo pulp_database`
+  `mongorestore --host localhost mongo_dump/pulp_database/`
+  `katello-service stop --only mongod`
+  puts "Done."
+end
+
+def valid_logical_backup
+  base_files_present =  @mongo_dump_exists && @mongo_data_exists
+  if @foreman_proxy_content
+    base_files_present
+  else
+    base_files_present && @pgsql_data_exists && @candlepin_dump_exists && @foreman_dump_exists
+  end
+end
+
+def valid_online_backup
+  @candlepin_dump_exists && @foreman_dump_exists && @mongo_dump_exists &&
+  !(@mongo_data_exists || @pgsql_data_exists)
+end
+
+def valid_fpc_online_backup
+  @mongo_dump_exists &&
+  !(@mongo_data_exists || @pgsql_data_exists || @candlepin_dump_exists || @foreman_dump_exists)
+end
+
+def valid_standard_backup
+  @mongo_data_exists && @pgsql_data_exists &&
+  !(@candlepin_dump_exists || @foreman_dump_exists || @mongo_dump_exists)
+end
+
+def valid_fpc_standard_backup
+  @mongo_data_exists &&
+  !(@pgsql_data_exists || @candlepin_dump_exists || @foreman_dump_exists || @mongo_dump_exists)
+end
+
+def backup_invalid_message
+  puts "There are too many files in the backup directory, please check only the " \
+       "required files are present."
+  display_backup_options
+  exit(-1)
+end
+
 def restore_databases
-  if File.exist?('pulp_data.tar')
+  puts "Logical backup detected, using the standard backup files to restore" if valid_logical_backup
+  if @pulp_data_exists
     puts "Restoring Pulp data"
     `tar --selinux --overwrite --listed-incremental=/dev/null -xf pulp_data.tar -C /`
   end
-  if File.exist?('mongo_data.tar.gz')
+  if @mongo_data_exists
     `tar --selinux --overwrite --listed-incremental=/dev/null -xzf mongo_data.tar.gz -C /`
+  end
+  if @pgsql_data_exists
     `tar --selinux --overwrite --listed-incremental=/dev/null -xzf pgsql_data.tar.gz -C /`
-  else
-    `service postgresql start`
-    `runuser - postgres -c "dropdb foreman"`
-    `runuser - postgres -c "dropdb candlepin"`
-    `runuser - postgres -c "psql -f #{DIR}/pg_globals.dump postgres 2>/dev/null"`
-    `runuser - postgres -c "pg_restore -C -d postgres #{DIR}/foreman.dump"`
-    `runuser - postgres -c "pg_restore -C -d postgres #{DIR}/candlepin.dump"`
-    `service postgresql stop`
-
-    `service mongod start`
-    `echo 'db.dropDatabase();' | mongo pulp_database`
-    `mongorestore --host localhost mongo_dump/pulp_database/`
-    `service mongod stop`
   end
+  if !@mongo_data_exists && !@pgsql_data_exists && !valid_logical_backup
+    if @foreman_dump_exists && @candlepin_dump_exists
+      restore_psql_dumps
+    end
+    if @mongo_dump_exists
+      restore_mongo_dump
+    end
+  end
+  migrate_pulp
```

Comment:
I've found that sometimes pulp needs to be migrated after a restore, especially if restoring from a different katello z-stream version. It won't change anything if pulp is already migrated,  but is there for extra insurance to make sure the pulp database is up to date after a restore